### ISSUE TITLE
Perl MNIST using record file format. Test not passed.

### ISSUE THREAD

**Issue by dokechin** [apache/incubator-mxnet#7511]
## Environment info
Operating System:Mac OS

Package used (Python/R/Scala/Julia):
Perl

MXNet version:
0.10.0

## Information
I am trying to use MNIST program, using mxnet record file format.
It seemed to data loading is going good, but accuracy test is not passed.
Source file is uploaded GitHub below.

https://github.com/dokechin/mnist_png

**Comment by sergeykolychev** [comment]
@dokechin Thanks! I'll review it on weekend and fix bugs if any.

**Comment by sergeykolychev** [comment]
@dokechin 
When I change your code to use ImageIter (and make batch size 100 and as well force shuffle => 1)
I get good testing and validation accuracy
```
  my $testing_ite = mx->img->ImageIter(
 {  batch_size => 100, data_shape=> [3,28,28],label_width =>1, path_imglist => "testing.ctl", path_root => '.'});
my $training_ite = mx->img->ImageIter(
 {  batch_size => 100, data_shape=> [3,28,28],label_width =>1, path_imglist => "training.ctl", path_root => '.', shuffle => 1 });
```
As well I commented out your optimizer_params and use optimizer 'adam';
```
$model->fit(
   $training_ite,
   eval_data => $testing_ite,
   optimizer => 'adam',
#   optimizer_params=>{learning_rate=>0.01},
   num_epoch=>4,
   batch_end_callback  => mx->callback->Speedometer(100, 10),
);
```
When I use the above optimizer setup I get good training accuracy even when using ImageRecordIter during one epoch though it gets reset every subsequent epoch and the validation accuracy is 10%
Something is fishy with our usage of the ImageRecordIter (btw it is not defined in the perl space and gets loaded directly from c++)
I never yet used ImageRecordIter so there are likely some bugs here, I'll try to get to the bottom of this.

**Comment by sergeykolychev** [comment]
@dokechin There's seems to be some error at C++ level, the training.ctl file that you used to create the training.rec contains 60,000 lines and the ImageRecordIter somehow outputs only data and labels for 9 (last class in the file), if I decrease number of lines in the file and recreate the .rec file the iterator starts working perfectly. There's seems to be some limit of the number of images.
This has no relation to the perl code I think. I'll try to reproduce it with python and if the bug stays I'll file the issue with the upstream.

**Comment by sergeykolychev** [comment]
@dokechin 
Found your problem
1) you used batch size 1 which is not recommended (for mnist 100 works well)
2) you must shuffle your data otherwise the network will see 5000 '9' number as last ones in the training and will forget all other digits
3) it's best to use 'adam' optimizer than the default 'sgd' because with it the network converges much faster
4) if you want to use your own original optimizer parameters and the default 'sgd' than increase number of epochs to say 4, 2 is not enough to get to 80% accuracy on the training set
5) The ImageRecIter needs a chunk of memory to shuffle your data, unfortunately if this memory size is insufficient some of the data will be returned unshuffled and this will prevent the network from converging
I put this number as 1000 (Mb) and it seems to be enough.
here is the patch to your perl script that passes the test:
```
developer@devbox:~/mnist_png$ git diff
diff --git a/image.pl b/image.pl
index f2549d2..c95eee6 100644
--- a/image.pl
+++ b/image.pl
@@ -5,11 +5,11 @@ use Data::Dumper;
 use Test::More tests => 1;
 
  my $testing_ite = mx->io->ImageRecordIter(
- {  batch_size => 1, data_shape=> [1,28,28],label_width =>1, path_imgrec => "testing.rec", path_root => '.' });
+ {  batch_size => 100, data_shape=> [1,28,28],label_width =>1, path_imgrec => "testing.rec", path_root => '.' });
 
 
  my $training_ite = mx->io->ImageRecordIter(
- {  batch_size => 1, data_shape=> [1,28,28],label_width =>1, path_imgrec => "training.rec", path_root => '.' });
+ {  batch_size => 100, data_shape=> [1,28,28],label_width =>1, path_imgrec => "training.rec", path_root => '.', shuffle => 1, shuffle_chunk_size => 1000 });
 
 
 # for my $data (@{$training_ite}){
@@ -44,7 +44,7 @@ my $model = mx->mod->Module(
 $model->fit(
    $training_ite,
    eval_data => $testing_ite,
-   optimizer_params=>{learning_rate=>0.01, momentum=> 0.9},
+   optimizer => 'adam',
    num_epoch=>2
 );
 my $res = $model->score($testing_ite, mx->metric->create('acc'));
```

**Comment by dokechin** [comment]
Thanks too much, I passed test!!
And I will try not MINIST but my own data.

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!
Also, do please check out our [forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) for general "how-to" questions.
